horace_core/sqw/@sqw/private/split_npix_full_bins.m

```diff
@@ -0,0 +1,83 @@
+function [npix_chunks, idxs] = split_npix_full_bins(npix, page_size, num_pix, npix_cum_sum)
+%SPLIT_NPIX_FULL_BINS Split the given array of counts such that the sum of each
+%  sub-array has a maximum of page_size.
+%
+% The npix counts are not split across bin boundaries - i.e. each sub-array
+% contains only full bins. If a value in npix is greater than page_size, then
+% that value will comprise its own sub-array.
+%
+% Input:
+% ------
+% npix          An array of integers (usually sqw.data.npix)
+% page_size     An integer specifying the maximum sum for each sub-array
+% num_pix       The sum of npix [optional].
+%               If this argument is not given it is calculated on-the-fly.
+%               In most cases, this value is cached for sqw objects in
+%               sqw.data.pix.num_pixels.
+% npix_cum_sum  The cumulative sum of the given npix array [optional].
+%               If not specified it will be calculated on-the-fly.
+%               Given as an optional argument as it's common to have already
+%               calculated this quantity (defines upper bounds of bin edges).
+%               This should be a vector.
+%
+% Output:
+% -------
+% npix_chunks  Cell array of vectors. Concatenation of these arrays will be
+%              equal to a flattened version the inputted npix array, i.e.
+%              npix(:).
+% idxs         The indices at which the input npix array was "split". Has size
+%              [2, n], where n is numel(npix_chunks). Each idxs(i, :) are the
+%              upper and lower indices into npix of npix_chunks{i}.
+%
+% Example:
+% --------
+% >> npix = [3, 2, 0, 6, 0, 5, 3, 1, 1, 4, 2, 3, 0];
+% >> page_size = 11;
+% >> [chunks, idxs] = split_npix_full_bins(npix, page_size)
+%   chunks =
+%       { [3, 2, 0, 6, 0], [5, 3, 1, 1], [4, 2, 3, 0] }
+%   idxs =
+%       1     6    10
+%       5     9    13
+%
+if ~exist('num_pix', 'var')
```

Comment:
Are there use-cases for `num_pix` and `npix_cum_sum` to be optional parameters? There's a lot of additional complexity introduced by potentially needing to calculate them

Reply:
It's just a relic of me originally deriving the values in this function, then realizing I already had them calculated where I was calling it.

Since this is a private SQW method and not a general "split this array of counts" utility, you're probably right, we can make the optional args compulsory. Unless we think this function is better as a utility - still not really sure where the best place for it is